File: lanqiao/15PythonA省赛/吊坠/吊坠.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <string>

using namespace std;

const int N = 205, M = 55;

int g[N][N], st[N], d[N];

string a[N];

int cal(string a, string b)
{
    int cnt = 0;
    int f[2 * M][2 * M];
    memset(f, 0, sizeof(f));

    // 从1开始，避免访问负索引
    for (int i = 1; i <= a.size(); i++)
    {
        for (int j = 1; j <= b.size(); j++)
        {
            if (a[i - 1] == b[j - 1]) // 对应字符相同，调整索引进行比较
            {
                f[i][j] = f[i - 1][j - 1] + 1;
                cnt = max(f[i][j], cnt);
            }
        }
    }
    return cnt;
}

int main()
{
    int n, m;
    cin >> n >> m;
    for (int i = 0; i < n; i++)
    {
        cin >> a[i];
        a[i] += a[i];
    }
    for (int i = 0; i < n; i++)
    {
        for (int j = i; j < n; j++)
        {
            g[i][j] = g[j][i] = min(m, cal(a[i], a[j]));
        }
    }
    // for (int i = 0; i < n; i++)
    // {
    //     for (int j = 0; j < n; j++)
    //     {
    //         cout << g[i][j] << ' ';
    //     }
    //     cout << endl;
    // }
    int res = 0;

    for (int i = 0; i < n; i++)
    {
        int t = -1;
        for (int j = 0; j < n; j++)
        {
            if (!st[j] && (t == -1 || d[t] < d[j]))
            {
                t = j;
            }
        }
        st[t] = 1;
        res += d[t];
        for (int j = 0; j < n; j++)
        {
            d[j] = max(d[j], g[t][j]);
        }
    }
    cout << res << endl;
    return 0;
}

File: lanqiao/15PythonA省赛/吊坠/吊坠.py
N = 205

n, m = map(int, input().strip().split())

a = []

g = [[0 for _ in range(N)] for _ in range(N)]
d = [float("-inf")] * N
st = [0] * N

for _ in range(n):
    a.append(input().strip())

for i in range(n):
    a[i] += a[i]

for i in range(n):

    sa = a[i]

    for j in range(i,n):
        f = [[0 for _ in range(2 * N)] for _ in range(2 * N)]
        sb = a[j]
        cnt = 0
        for k in range(2 * m):
            for z in range(2 * m):
                if (sa[k] == sb[z]):
                    f[k][z] = max(f[k - 1][z - 1] + 1, f[k][z])
                    cnt = max(cnt, f[k][z])
        g[i][j] = min(m, cnt)
        g[j][i] = min(m, cnt)

d[0] = 0

res = 0
for i in range(n):
    t = - 1
    for j in range(n):
        if st[j] == 0 and (t == - 1 or d[t] < d[j]):
            t = j
    st[t] = 1
    res += d[t]

    for j in range(n):
        d[j] = max(d[j], g[t][j])

print(res)
